File: CommandPattern(Behavioral)/CommandPatternWithHistory/undo_redo_java/HistoryList.java
import java.util.ArrayList;
import java.util.List;

public class HistoryList {
    private final List<Command> commandList = new ArrayList<>();
    private final List<Command> undoList = new ArrayList<>();

    public void undo() {
        if (!commandList.isEmpty()) {
            Command lastCommand = commandList.removeLast();
            lastCommand.unExecute();
            undoList.add(lastCommand);
        }
    }

    public void redo() {
        // if undo list is empty it will do nothing
        if (!undoList.isEmpty()) {
            Command lastCommand = undoList.removeLast();
            lastCommand.execute();
            commandList.add(lastCommand);
        }
    }

    public void addCommand(Command command) {
        commandList.add(command);
    }
}

File: Behavioral/CommandPattern(Behavioral)/CommandPatternWithHistory/undo_redo/HistoryList.ts
import { Command } from "./Command";

export class HistoryList {
  private commandList: Command[] = [];
  private undoList: Command[] = [];

  public undo(): void {
    if (this.commandList.length > 0) {
      const commandObj: Command | undefined = this.commandList.pop();
      if (commandObj !== undefined) {
        commandObj.unExecute();
        this.undoList.push(commandObj);
      }
    }
  }

  public redo(): void {
    if (this.undoList.length > 0) {
      const commandObj: Command | undefined = this.undoList.pop();
      if (commandObj !== undefined) {
        commandObj.execute();
        this.commandList.push(commandObj);
      }
    }
  }

  public addCommand(command: Command): void {
    this.commandList.push(command);
  }
}
